### PR TITLE
Initial event-sourced state machine

### PR DESCRIPTION
**Pull request by jessemyers**
Includes:

 -  An abstraction for defining event type enumerations and their behavior
 -  A meta class for defining new event models
 -  A base store for common event persistence
 -  A base controller for common event business logic
 -  A starting place for resources and routes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,133 @@
 # microcosm-eventsource
-Event sourced state machines
+
+Event-sourced state machines using `microcosm`.
```
Comment (jessemyers) [inline README.md]: Start here...

File: README.md [hunk 2]
```diff
@@ -1,2 +1,133 @@
...
+ -  After a task is started, it may be reassigned or rescheduled any number of times.
+
+ -  After a task is started, it may either be canceled or completed. Both of these
+    states are terminal.
+
+ -  Before a task is canceled or completed, it maybe revised at any time. A revised
+    task reverts back to the initial state.
+
+
+These actions collectively define a state machine, where each transition is triggered
```
Comment (jessemyers) [inline README.md]: My diagram is missing a few states, but you get the idea.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,133 @@
...
+ -  Each enumerated value is configured using its (left hand side) meta data.
+ -  This meta data includes:
+     -  "following conditions" for legal state transitions.
+     -  a flag controlling whether state accumulates through this transition
+     -  a flag controlling whether the state machine is restarted (with a new version) after this state
+     -  a list of zero or more required additional columns for this state
+
+Example:
+
+    class TaskEventType(EventType):
```
Comment (jessemyers) [inline README.md]: The most significant changes are all in how we defined these enumerations.

File: microcosm_eventsource/controllers.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        self.validate_required_fields(event_type, **kwargs)
+        parent = self.store.retrieve_most_recent(**kwargs)
+        version, state = self.process_state_transition(parent, event_type)
+        event = self.create_event(parent, event_type, version, state, **kwargs)
+        self.publish_event(event)
+        return event
+
+    def validate_required_fields(self, event_type, **kwargs):
+        """
+        Validate type-specific required fields.
```
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: Inspect the required fields for event types and raise 422 before the DB catches the constraint violation (and raises 500)

File: microcosm_eventsource/controllers.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                                event_type.name,
+                                camelize(required_field, uppercase_first_letter=False),
+                            )
+                        ],
+                    }
+                    for required_field in missing_required_fields
+                ],
+            )
+
+    def process_state_transition(self, parent, event_type):
```
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: Process a state transition by:
 - Checking the DNF syntax to determine if the transition is legal
 - Determining the state to transition to
 - Determine the version to transition to

File: microcosm_eventsource/controllers.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                raise IllegalStateTransitionError("Event type '{}' may not follow [{}]".format(
+                    event_type.name,
+                    ", ".join(item.name for item in parent.state),
+                ))
+
+        parent_version = parent.version if parent else 1
+        parent_state = parent.state if parent else []
+
+        if event_type.is_accumulating:
+            state = parent_state + [event_type]
```
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: I need better tests for this logic; we probably don't want to repeat event types during accumulation.

File: microcosm_eventsource/controllers.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            version = parent_version
+
+        return version, state
+
+    def create_event(self, parent, event_type, version, state, **kwargs):
+        """
+        Create the event.
+
+        """
+        # XXX we should upsert here as long as we can reason about the constraints
```
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: There are race conditions we can solve with upsert, but it requires a bit more care with our unique constraints.

File: microcosm_eventsource/models.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        return (
+            # logical clock is unique and indexed
+            Index(
+                "unique_logical_clock",
+                container_id_name,
+                "clock",
+                unique=True,
+            ),
+            # events are unique per container, event type, and version
+            # XXX this won't work for some types of events
```
Comment (jessemyers) [inline microcosm_eventsource/models.py]: For my fixture example, declaring uniqueness by event type doesn't work for reassign/reschedule states. To do.

File: microcosm_eventsource/models.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            CheckConstraint(
+                name="require_parent_id",
+                sqltext="parent_id IS NOT NULL OR (version = 1 AND event_type IN ({}))".format(
+                    join_event_types(item for item in event_type if item.is_initial),
+                ),
+            ),
+        )
+
+    def make_column_constraints(cls, event_type):
+        return (
```
Comment (jessemyers) [inline microcosm_eventsource/models.py]: The `assignee` and `deadline` attributes need to be generalized.

File: microcosm_eventsource/tests/fixtures.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        requires=["assignee"],
+    )
+    RESCHEDULED = info(
+        follows=["STARTED"],
+        accumulating=True,
+        requires=["deadline"],
+    )
+    REVISED = info(
+        follows=["CREATED", "STARTED"],
+        # XXX need to support jumping to state CREATED
```
Comment (jessemyers) [inline microcosm_eventsource/tests/fixtures.py]: Two choices here:
 - Let event types have more control over their state transition logic (probably)
 - Write the state machine to handle `REVISED` or `CREATED` everywhere.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,133 @@
...
+        def __init__(self, graph):
+            super(TaskEventStore, self).__init__(graph, TaskEvent)
+
+
+### Defining Task Event Types
+
+It remains to define the legal events and their transitions using an enumeration. This is done
+as follows:
+
+ -  Each enumerated value is identified using its (right hand side) `name`.
```
Comment (netama) [inline README.md]: swapped sides?

File: README.md [hunk 12]
```diff
@@ -1,2 +1,133 @@
...
+     -  a list of zero or more required additional columns for this state
+
+Example:
+
+    class TaskEventType(EventType):
+        CREATED = info()
+        ASSIGNED = info(
+            follows=[all_of("CREATED", but_not("ASSIGNED"))],
+            accumulating=True,
+            requires=["assignee"]
```
Comment (netama) [inline README.md]: nit - missing comma

File: microcosm_eventsource/controllers.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    IllegalInitialStateError,
+    IllegalStateTransitionError,
+)
+
+
+class EventController(CRUDStoreAdapter):
+
+    @property
+    def sns_producer(self):
+        return self.graph.sns_producer
```
Comment (netama) [inline microcosm_eventsource/controllers.py]: is the graph always available now?

File: microcosm_eventsource/controllers.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        parent_version = parent.version if parent else 1
+        parent_state = parent.state if parent else []
+
+        if event_type.is_accumulating:
+            state = parent_state + [event_type]
+        else:
+            state = [event_type]
+
+        if event_type.is_restarting:
+            version = parent_version + 1
```
Comment (netama) [inline microcosm_eventsource/controllers.py]: does this mean the initial event can't be restarting? (if it can, I think this might end up being 2 for the initial event)

File: microcosm_eventsource/event_types.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+An event type defines a unique event and the logic for evaluating state transitions using
+a DNF mini grammar:
+
+ -  An event can *follow* a disjuction of following conditions
+ -  A following condition can match an event type exactly
+ -  A following condition can match a conjunction of following conditions
+ -  A following condition can match the negation of a following conditions
+
+This matching logic depends is applied again *state* accumulation. By default, event state
```
Comment (netama) [inline microcosm_eventsource/event_types.py]: can you rephrase the first sentence?

**Comment by netama** [comment]
LGTM with a few comments

File: microcosm_eventsource/controllers.py (hunk 13 above)
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: In this particular base class...

File: microcosm_eventsource/controllers.py (hunk 14 above)
Comment (jessemyers) [inline microcosm_eventsource/controllers.py]: Will test.
